File: python/song/buffer.py
# MIT License
# Copyright (c) 2026 dbjwhs

"""
Buffer class for Song wire protocol serialization.

All integers are little-endian to match the C++ implementation.
"""

import struct
from typing import List, Optional, Callable, TypeVar

T = TypeVar('T')


class BufferError(Exception):
    """Raised when buffer operations fail."""
    pass


class Buffer:
    """
    A byte buffer for serializing and deserializing Song wire protocol data.

    Matches the C++ Buffer class wire format exactly.
    """

    def __init__(self, data: bytes = b""):
        self._data = bytearray(data)
        self._read_pos = 0

    # =========================================================================
    # Basic Operations
    # =========================================================================

    def write(self, data: bytes) -> None:
        """Write raw bytes to the buffer."""
        self._data.extend(data)

    def read(self, length: int) -> bytes:
        """Read raw bytes from the buffer."""
        if self._read_pos + length > len(self._data):
            raise BufferError(
                f"Buffer underflow: attempted to read {length} bytes "
                f"but only {len(self._data) - self._read_pos} available"
            )
        result = bytes(self._data[self._read_pos:self._read_pos + length])
        self._read_pos += length
        return result

    def data(self) -> bytes:
        """Get all data in the buffer."""
        return bytes(self._data)

    def size(self) -> int:
        """Get current buffer size."""
        return len(self._data)

    def remaining(self) -> int:
        """Get remaining bytes to read."""
        return len(self._data) - self._read_pos

    def has_data(self) -> bool:
        """Check if more data available to read."""
        return self._read_pos < len(self._data)

    def reset(self) -> None:
        """Reset buffer (clear all data)."""
        self._data = bytearray()
        self._read_pos = 0

    def reset_read(self) -> None:
        """Reset read position to beginning."""
        self._read_pos = 0

    # =========================================================================
    # Primitive Encoding (Little Endian)
    # =========================================================================

    def encode_bool(self, val: bool) -> None:
        """Encode a boolean (1 byte)."""
        self._data.append(1 if val else 0)

    def encode_i8(self, val: int) -> None:
        """Encode a signed 8-bit integer."""
        self._data.extend(struct.pack('<b', val))

    def encode_i16(self, val: int) -> None:
        """Encode a signed 16-bit integer."""
        self._data.extend(struct.pack('<h', val))

    def encode_i32(self, val: int) -> None:
        """Encode a signed 32-bit integer."""
        self._data.extend(struct.pack('<i', val))

    def encode_i64(self, val: int) -> None:
        """Encode a signed 64-bit integer."""
        self._data.extend(struct.pack('<q', val))

    def encode_u8(self, val: int) -> None:
        """Encode an unsigned 8-bit integer."""
        self._data.extend(struct.pack('<B', val))

    def encode_u16(self, val: int) -> None:
        """Encode an unsigned 16-bit integer."""
        self._data.extend(struct.pack('<H', val))

    def encode_u32(self, val: int) -> None:
        """Encode an unsigned 32-bit integer."""
        self._data.extend(struct.pack('<I', val))

    def encode_u64(self, val: int) -> None:
        """Encode an unsigned 64-bit integer."""
        self._data.extend(struct.pack('<Q', val))

    def encode_f32(self, val: float) -> None:
        """Encode a 32-bit float."""
        self._data.extend(struct.pack('<f', val))

    def encode_f64(self, val: float) -> None:
        """Encode a 64-bit float."""
        self._data.extend(struct.pack('<d', val))

    def encode_string(self, val: str) -> None:
        """Encode a string (length-prefixed UTF-8)."""
        encoded = val.encode('utf-8')
        if len(encoded) > 0xFFFFFFFF:
            raise BufferError("String too large to encode")
        self.encode_u32(len(encoded))
        self._data.extend(encoded)

    def encode_bytes(self, val: bytes) -> None:
        """Encode raw bytes (length-prefixed)."""
        if len(val) > 0xFFFFFFFF:
            raise BufferError("Bytes too large to encode")
        self.encode_u32(len(val))
        self._data.extend(val)

    # =========================================================================
    # Primitive Decoding (Little Endian)
    # =========================================================================

    def decode_bool(self) -> bool:
        """Decode a boolean."""
        return self.read(1)[0] != 0

    def decode_i8(self) -> int:
        """Decode a signed 8-bit integer."""
        return struct.unpack('<b', self.read(1))[0]

    def decode_i16(self) -> int:
        """Decode a signed 16-bit integer."""
        return struct.unpack('<h', self.read(2))[0]

    def decode_i32(self) -> int:
        """Decode a signed 32-bit integer."""
        return struct.unpack('<i', self.read(4))[0]

    def decode_i64(self) -> int:
        """Decode a signed 64-bit integer."""
        return struct.unpack('<q', self.read(8))[0]

    def decode_u8(self) -> int:
        """Decode an unsigned 8-bit integer."""
        return struct.unpack('<B', self.read(1))[0]

    def decode_u16(self) -> int:
        """Decode an unsigned 16-bit integer."""
        return struct.unpack('<H', self.read(2))[0]

    def decode_u32(self) -> int:
        """Decode an unsigned 32-bit integer."""
        return struct.unpack('<I', self.read(4))[0]

    def decode_u64(self) -> int:
        """Decode an unsigned 64-bit integer."""
        return struct.unpack('<Q', self.read(8))[0]

    def decode_f32(self) -> float:
        """Decode a 32-bit float."""
        return struct.unpack('<f', self.read(4))[0]

    def decode_f64(self) -> float:
        """Decode a 64-bit float."""
        return struct.unpack('<d', self.read(8))[0]

    def decode_string(self) -> str:
        """Decode a string."""
        length = self.decode_u32()
        return self.read(length).decode('utf-8')

    def decode_bytes(self) -> bytes:
        """Decode raw bytes."""
        length = self.decode_u32()
        return self.read(length)

    # =========================================================================
    # Array Encoding/Decoding
    # =========================================================================

    def encode_array_i32(self, arr: List[int]) -> None:
        """Encode an array of i32."""
        self.encode_u32(len(arr))
        for val in arr:
            self.encode_i32(val)

    def decode_array_i32(self) -> List[int]:
        """Decode an array of i32."""
        count = self.decode_u32()
        return [self.decode_i32() for _ in range(count)]

    def encode_array_i64(self, arr: List[int]) -> None:
        """Encode an array of i64."""
        self.encode_u32(len(arr))
        for val in arr:
            self.encode_i64(val)

    def decode_array_i64(self) -> List[int]:
        """Decode an array of i64."""
        count = self.decode_u32()
        return [self.decode_i64() for _ in range(count)]

    def encode_array_f32(self, arr: List[float]) -> None:
        """Encode an array of f32."""
        self.encode_u32(len(arr))
        for val in arr:
            self.encode_f32(val)

    def decode_array_f32(self) -> List[float]:
        """Decode an array of f32."""
        count = self.decode_u32()
        return [self.decode_f32() for _ in range(count)]

    def encode_array_f64(self, arr: List[float]) -> None:
        """Encode an array of f64."""
        self.encode_u32(len(arr))
        for val in arr:
            self.encode_f64(val)

    def decode_array_f64(self) -> List[float]:
        """Decode an array of f64."""
        count = self.decode_u32()
        return [self.decode_f64() for _ in range(count)]

    def encode_array_string(self, arr: List[str]) -> None:
        """Encode an array of strings."""
        self.encode_u32(len(arr))
        for val in arr:
            self.encode_string(val)

    def decode_array_string(self) -> List[str]:
        """Decode an array of strings."""
        count = self.decode_u32()
        return [self.decode_string() for _ in range(count)]

    def encode_array(self, arr: List[T], encoder: Callable[['Buffer', T], None]) -> None:
        """Encode an array using a custom encoder function."""
        self.encode_u32(len(arr))
        for val in arr:
            encoder(self, val)

    def decode_array(self, decoder: Callable[['Buffer'], T]) -> List[T]:
        """Decode an array using a custom decoder function."""
        count = self.decode_u32()
        return [decoder(self) for _ in range(count)]

    # =========================================================================
    # Optional Encoding/Decoding
    # =========================================================================

    def encode_optional(self, val: Optional[T], encoder: Callable[['Buffer', T], None]) -> None:
        """Encode an optional value."""
        if val is None:
            self.encode_bool(False)
        else:
            self.encode_bool(True)
            encoder(self, val)

    def decode_optional(self, decoder: Callable[['Buffer'], T]) -> Optional[T]:
        """Decode an optional value."""
        has_value = self.decode_bool()
        if has_value:
            return decoder(self)
        return None

File: src/buffer.cpp
// MIT License
// Copyright (c) 2026 dbjwhs

#include "song/buffer.hpp"
#include <stdexcept>
#include <limits>
#include <bit>

// Song specifies a little-endian wire format. All numeric types are serialized
// via memcpy without byte swapping, so both endpoints must be little-endian.
// This covers x86, ARM (LE mode), and RISC-V — effectively all modern targets.
static_assert(std::endian::native == std::endian::little,
              "Song wire protocol requires a little-endian architecture");

namespace song {

Buffer::~Buffer() {
    if (!is_inline()) {
        delete[] data_;
    }
}

Buffer::Buffer(Buffer&& other) noexcept
    : size_(other.size_), capacity_(other.capacity_), read_pos_(other.read_pos_) {
    if (other.is_inline()) {
        // Only copy the bytes actually in use, not the full 4KB buffer
        std::memcpy(inline_, other.inline_, other.size_);
        data_ = inline_;
    } else {
        data_ = other.data_;
        other.data_ = other.inline_;
        other.capacity_ = kInlineSize;
    }
    other.size_ = 0;
    other.read_pos_ = 0;
}

Buffer& Buffer::operator=(Buffer&& other) noexcept {
    if (this != &other) {
        if (!is_inline()) {
            delete[] data_;
        }

        size_ = other.size_;
        capacity_ = other.capacity_;
        read_pos_ = other.read_pos_;

        if (other.is_inline()) {
            // Only copy the bytes actually in use, not the full 4KB buffer
            std::memcpy(inline_, other.inline_, other.size_);
            data_ = inline_;
        } else {
            data_ = other.data_;
            other.data_ = other.inline_;
            other.capacity_ = kInlineSize;
        }
        other.size_ = 0;
        other.read_pos_ = 0;
    }
    return *this;
}

void Buffer::ensure_capacity(size_t required) {
    if (required <= capacity_) return;

    // Guard against overflow: capacity doubling must not wrap around SIZE_MAX
    constexpr size_t kMaxCapacity = std::numeric_limits<size_t>::max() / 2;
    if (required > kMaxCapacity) {
        throw std::runtime_error("Buffer capacity overflow: requested size too large");
    }

    size_t new_capacity = capacity_;
    while (new_capacity < required) {
        new_capacity *= 2;
    }

    auto* new_data = new std::byte[new_capacity];
    std::memcpy(new_data, data_, size_);

    if (!is_inline()) {
        delete[] data_;
    }

    data_ = new_data;
    capacity_ = new_capacity;
}

void Buffer::write(const void* data, size_t len) {
    if (len > 0 && size_ > std::numeric_limits<size_t>::max() - len) {
        throw std::runtime_error("Buffer write overflow: size + len exceeds maximum");
    }
    ensure_capacity(size_ + len);
    std::memcpy(data_ + size_, data, len);
    size_ += len;
}

void Buffer::read(void* data, size_t len) {
    if (len > size_ || read_pos_ > size_ - len) {
        throw std::runtime_error("Buffer underflow: attempted to read beyond buffer size");
    }
    std::memcpy(data, data_ + read_pos_, len);
    read_pos_ += len;
}

// Primitive encoders
void encode_bool(Buffer& buf, bool val) {
    u8 byte = val ? 1 : 0;
    buf.write(&byte, 1);
}

void encode_i8(Buffer& buf, i8 val) {
    buf.write(&val, sizeof(val));
}

void encode_i16(Buffer& buf, i16 val) {
    buf.write(&val, sizeof(val));
}

void encode_i32(Buffer& buf, i32 val) {
    buf.write(&val, sizeof(val));
}

void encode_i64(Buffer& buf, i64 val) {
    buf.write(&val, sizeof(val));
}

void encode_u8(Buffer& buf, u8 val) {
    buf.write(&val, sizeof(val));
}

void encode_u16(Buffer& buf, u16 val) {
    buf.write(&val, sizeof(val));
}

void encode_u32(Buffer& buf, u32 val) {
    buf.write(&val, sizeof(val));
}

void encode_u64(Buffer& buf, u64 val) {
    buf.write(&val, sizeof(val));
}

void encode_f32(Buffer& buf, f32 val) {
    buf.write(&val, sizeof(val));
}

void encode_f64(Buffer& buf, f64 val) {
    buf.write(&val, sizeof(val));
}

void encode_string(Buffer& buf, std::string_view val) {
    if (val.size() > kMaxStringSize) {
        throw std::runtime_error("String too large to encode");
    }
    u32 len = static_cast<u32>(val.size());
    encode_u32(buf, len);
    buf.write(val.data(), len);
}

void encode_bytes(Buffer& buf, std::span<const std::byte> val) {
    if (val.size() > kMaxBytesSize) {
        throw std::runtime_error("Byte array too large to encode");
    }
    u32 len = static_cast<u32>(val.size());
    encode_u32(buf, len);
    buf.write(val.data(), len);
}

// Primitive decoders
bool decode_bool(Buffer& buf) {
    u8 byte;
    buf.read(&byte, 1);
    return byte != 0;
}

i8 decode_i8(Buffer& buf) {
    i8 val;
    buf.read(&val, sizeof(val));
    return val;
}

i16 decode_i16(Buffer& buf) {
    i16 val;
    buf.read(&val, sizeof(val));
    return val;
}

i32 decode_i32(Buffer& buf) {
    i32 val;
    buf.read(&val, sizeof(val));
    return val;
}

i64 decode_i64(Buffer& buf) {
    i64 val;
    buf.read(&val, sizeof(val));
    return val;
}

u8 decode_u8(Buffer& buf) {
    u8 val;
    buf.read(&val, sizeof(val));
    return val;
}

u16 decode_u16(Buffer& buf) {
    u16 val;
    buf.read(&val, sizeof(val));
    return val;
}

u32 decode_u32(Buffer& buf) {
    u32 val;
    buf.read(&val, sizeof(val));
    return val;
}

u64 decode_u64(Buffer& buf) {
    u64 val;
    buf.read(&val, sizeof(val));
    return val;
}

f32 decode_f32(Buffer& buf) {
    f32 val;
    buf.read(&val, sizeof(val));
    return val;
}

f64 decode_f64(Buffer& buf) {
    f64 val;
    buf.read(&val, sizeof(val));
    return val;
}

std::string decode_string(Buffer& buf) {
    u32 len = decode_u32(buf);
    if (len > kMaxStringSize) {
        throw std::runtime_error("String too large");
    }
    std::string result(len, '\0');
    buf.read(result.data(), len);
    return result;
}

std::vector<std::byte> decode_bytes(Buffer& buf) {
    u32 len = decode_u32(buf);
    if (len > kMaxBytesSize) {
        throw std::runtime_error("Byte array too large");
    }
    std::vector<std::byte> result(len);
    buf.read(result.data(), len);
    return result;
}

} // namespace song
